### PR TITLE
C#: ZipSlip - Refine sanitizers

### PR DESCRIPTION
**Pull request by lukecartey**
1. Treat the `GetFileName` and `Substring` _calls_ as sanitizers, rather than the arguments. This is because it is the _result_ of the calls `GetFileName` and `Substring` which should be considered sanitized. By treating the argument as sanitized, we could erroneously suppress some use-use flow. For example:
```
var path = Path.Combine(destDir, entry.FullName);
var filename = Path.GetFileName(path);
log("Extracting " + filename);
entry.ExtractToFile(path);
```
2. ZipSlip can be avoided by checking that the combined and resolved path `StartsWith` the appropriate destination directory. Refine the `StartsWith` sanitizer to:
     * Consider expressions _guarded_ by an appropriate `StartsWith` check to be sanitized.
     * Consider a `StartsWith` check to be inappropriate if it is checking the result of `Path.Combine`, as that has not been appropriately resolved.

@denislevin fyi

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ZipSlip.qll [hunk 1]
```diff
@@ -96,31 +98,55 @@ module ZipSlip {
...
         mc.getTarget().hasQualifiedName("System.IO.Path", "GetFileName") |
-        this.asExpr() = mc.getAnArgument()
+        this.asExpr() = mc
       )
     }
   }
 
   /**
-   * A qualifier in a call to `StartsWith` or `Substring` string method.
+   * A call to Substring.
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ZipSlip.qll]: Backticks

File: csharp/ql/src/semmle/code/csharp/security/dataflow/ZipSlip.qll [hunk 2]
```diff
@@ -96,31 +98,55 @@ module ZipSlip {
...
    *
-   * A call to a String method such as `StartsWith` or `Substring` can indicate a check for a
-   * relative path, or a check against the destination folder for whitelisted/target path, etc.
+   * This is considered a sanitizer because `Substring` may be used to extract a single component
+   * of a path to avoid ZipSlip.
    */
-  class StringCheckSanitizer extends Sanitizer {
-    StringCheckSanitizer() {
+  class SubstringSanitizer extends Sanitizer {
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/security/dataflow/ZipSlip.qll]: Could you add a test for this as well?

Comment (lukecartey) [reply csharp/ql/src/semmle/code/csharp/security/dataflow/ZipSlip.qll]: Done.

**Comment by lukecartey** [comment]
Comments addressed, changes pushed.

**Pull request merged**